src/hotspot/share/runtime/jniHandles.cpp

```diff
@@ -384,10 +384,15 @@ void JNIHandleBlock::release_block(JNIHandleBlock* block, JavaThread* thread) {
       block->_next = freelist;
     }
     block = NULL;
-  }
-  if (block != NULL) {
-    Atomic::dec(&_blocks_allocated);
-    delete block;
+  } else {
+    DEBUG_ONLY(block->set_pop_frame_link(NULL));
+    while (block != NULL) {
+      JNIHandleBlock* next = block->_next;
+      Atomic::dec(&_blocks_allocated);
+      assert(block->pop_frame_link() == NULL, "pop_frame_link should be NULL");
+      delete block;
+      block = next;
+    }
```

Comment:
Given the comment below about the `pop_frame_link`, I'm concerned that it may be possible that the loop above has deleted the block that `pop_frame_link` refers to. AFAICS we should only have a chain above if we have unbalanced `pushLocalFrame`/`popLocalFrame`, but that is the case the following code is trying to deal with too.